server/store/memory/peer_store.go

```diff
@@ -280,3 +289,51 @@ func (s *peerStore) AnnouncePeers(infoHash chihaya.InfoHash, seeder bool, numWan
 
 	return
 }
+
+func (s *peerStore) GetSeeders(infoHash chihaya.InfoHash) (peers, peers6 []chihaya.Peer, err error) {
+	skey := seedersKey(infoHash)
+
+	shard := s.shards[s.shardIndex(infoHash)]
+	shard.RLock()
+	defer shard.RUnlock()
+
+	seeders := shard.peers[skey]
+	for _, p := range seeders {
+		if p.IP.To4() == nil {
+			peers6 = append(peers6, p.Peer)
+		} else {
+			peers = append(peers, p.Peer)
+		}
+	}
+	return
+}
+
+func (s *peerStore) GetLeechers(infoHash chihaya.InfoHash) (peers, peers6 []chihaya.Peer, err error) {
+	lkey := leechersKey(infoHash)
+
+	shard := s.shards[s.shardIndex(infoHash)]
+	shard.RLock()
+	defer shard.RUnlock()
+
+	leechers := shard.peers[lkey]
+	for _, p := range leechers {
+		if p.IP.To4() == nil {
+			peers6 = append(peers6, p.Peer)
+		} else {
+			peers = append(peers, p.Peer)
+		}
+	}
+	return
+}
+
+func (s *peerStore) NumSeeders(infoHash chihaya.InfoHash) (uint32, error) {
+	seeders, seeders6, err := s.GetSeeders(infoHash)
+	seeders = append(seeders, seeders6...)
```

Comment:
Implement `NumSeeders` like this:

```go
skey := seedersKey(infoHash)

shard := s.shards[s.shardIndex(infoHash)]
shard.RLock()
defer shard.RUnlock()

return len(shard.peers[skey])
```

Although I don't know how performant `len` calls on maps are - I have a repo about microbenchmarks, I should add that.

I thought the idea behind `Get(Seeder|Leechers)` was to get all seeders/leechers so that the middleware can figure out which ones it wants?